src/audio_core/hle/wmf_decoder.cpp

```diff
@@ -0,0 +1,254 @@
+// Copyright 2018 Citra Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "audio_core/hle/wmf_decoder.h"
+#include "audio_core/hle/wmf_decoder_utils.h"
+
+namespace AudioCore::HLE {
+
+class WMFDecoder::Impl {
+public:
+    explicit Impl(Memory::MemorySystem& memory);
+    ~Impl();
+    std::optional<BinaryResponse> ProcessRequest(const BinaryRequest& request);
+
+private:
+    std::optional<BinaryResponse> Initalize(const BinaryRequest& request);
+
+    void Clear();
+
+    std::optional<BinaryResponse> Decode(const BinaryRequest& request);
+
+    int DecodingLoop(ADTSData adts_header, std::array<std::vector<u8>, 2>& out_streams);
+
+    bool initalized = false;
+    bool selected = false;
+
+    Memory::MemorySystem& memory;
+
+    IMFTransform* transform = NULL;
+    DWORD in_stream_id = 0;
+    DWORD out_stream_id = 0;
+};
+
+WMFDecoder::Impl::Impl(Memory::MemorySystem& memory) : memory(memory) {
+    mf_coinit();
+}
+
+WMFDecoder::Impl::~Impl() = default;
+
+std::optional<BinaryResponse> WMFDecoder::Impl::ProcessRequest(const BinaryRequest& request) {
+    if (request.codec != DecoderCodec::AAC) {
+        LOG_ERROR(Audio_DSP, "Got unknown codec {}", static_cast<u16>(request.codec));
+        return {};
+    }
+
+    switch (request.cmd) {
+    case DecoderCommand::Init: {
+        LOG_INFO(Audio_DSP, "AACDecoder initializing");
+        return Initalize(request);
+    }
+    case DecoderCommand::Decode: {
+        return Decode(request);
+    }
+    case DecoderCommand::Unknown: {
+        BinaryResponse response;
+        std::memcpy(&response, &request, sizeof(response));
+        response.unknown1 = 0x0;
+        return response;
+    }
+    default:
+        LOG_ERROR(Audio_DSP, "Got unknown binary request: {}", static_cast<u16>(request.cmd));
+        return {};
+    }
+}
+
+std::optional<BinaryResponse> WMFDecoder::Impl::Initalize(const BinaryRequest& request) {
+    if (initalized) {
+        Clear();
+    }
+
+    BinaryResponse response;
+    std::memcpy(&response, &request, sizeof(response));
+    response.unknown1 = 0x0;
+
+    if (mf_decoder_init(&transform) != 0) {
+        LOG_CRITICAL(Audio_DSP, "Can't init decoder");
+        return response;
+    }
+
+    HRESULT hr = transform->GetStreamIDs(1, &in_stream_id, 1, &out_stream_id);
+    if (hr == E_NOTIMPL) {
+        // if not implemented, it means this MFT does not assign stream ID for you
+        in_stream_id = 0;
+        out_stream_id = 0;
+    } else if (FAILED(hr)) {
+        ReportError("Decoder failed to initialize the stream ID", hr);
+        SafeRelease(&transform);
+        return response;
+    }
+
+    initalized = true;
+    return response;
+}
+
+void WMFDecoder::Impl::Clear() {
+    if (initalized) {
+        mf_flush(&transform);
+        mf_deinit(&transform);
+    }
+    initalized = false;
+    selected = false;
+}
+
+int WMFDecoder::Impl::DecodingLoop(ADTSData adts_header,
```

Comment:
`std::optional<std::array<std::vector<u8>, 2>> WMFDecoder::Impl::DecodingLoop(const ADTSData& adts_header)`

Reply:
So how do I check if the `DecodingLoop` succeeded w/o status code? It's not like MFT will always return something if it returns `S_OK`

Reply:
with the optional. if optional has a value it succeed if not it failed. You never check what the failed error code is, nevertheless

Reply:
>with the optional. if optional has a value it succeed if not it failed. You never check what the failed error code is, nevertheless

Also, you should always check the error code __despite__ if there is any output you received. This is a strange behavior I have observed. When MFT returned a `MORE_DATA_AVAILABLE`, sometimes you will get a `NULL` next time you try to receive the sample and MFT will tell you the operation succeeded with 0 length of data.